### PR TITLE
Move InstKinds out to their own vector.

### PR DESCRIPTION
**Pull request by jonmeow**
This is intended to make InstKind consume a quarter the current space (1 byte instead of 4). The expectation is this'll be roughly performance neutral (but currently looks slightly negative), although we can start changing things to enable better performance for things that are happy to only examine the InstKind.

For a file generated by `source_gen`, before (with the mem usage fix):

```
sem_ir_.insts_.loc_ids_:
  used_bytes:      34694064
  reserved_bytes:  54525948
sem_ir_.insts_.values_:
  used_bytes:      138776256
  reserved_bytes:  218103792
```

After:

```
sem_ir_.insts_.loc_ids_:
  used_bytes:      34694064
  reserved_bytes:  54525948
sem_ir_.insts_.kinds_:
  used_bytes:      8673516
  reserved_bytes:  10747903
sem_ir_.insts_.types_and_args_:
  used_bytes:      104082192
  reserved_bytes:  163577844
...
Total:
  used_bytes:      493221460
  reserved_bytes:  650325525
```

And for `bazel run -c opt :compile_benchmark -- --benchmark_filter=Check`, before:

```
----------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations      Bytes      Lines     Tokens
----------------------------------------------------------------------------------------------------------------------------
BM_CompileAPIFileDenseDecls<Phase::Check>/256       1057130 ns      1056661 ns         1024 4.98457M/s 184.543k/s 1.04196M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/1024      1960553 ns      1959331 ns         1024 16.5817M/s  499.66k/s  2.9357M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/4096      5503917 ns      5500405 ns          256  25.795M/s  732.31k/s 4.33768M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/16384    19112623 ns     19102923 ns           64  31.982M/s 853.953k/s 5.06682M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/65536    79289121 ns     79233317 ns           16 31.8608M/s 826.168k/s 4.90462M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/262144  346024763 ns    345771854 ns            8 29.5566M/s 758.098k/s 4.50093M/s
```

Without `InstStore::Is` changes:

```
----------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations      Bytes      Lines     Tokens
----------------------------------------------------------------------------------------------------------------------------
BM_CompileAPIFileDenseDecls<Phase::Check>/256       1079970 ns      1079691 ns         1024 4.87825M/s 180.607k/s 1.01974M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/1024      2028099 ns      2027076 ns         1024 16.0275M/s 482.962k/s 2.83758M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/4096      5584768 ns      5581419 ns          256 25.4206M/s  721.68k/s 4.27472M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/16384    20294882 ns     20283454 ns           64 30.1206M/s 804.252k/s 4.77192M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/65536    83486939 ns     83452746 ns           16 30.2499M/s 784.396k/s 4.65664M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/262144  351386153 ns    351165409 ns            8 29.1027M/s 746.455k/s  4.4318M/s
```

With `InstStore::Is` changes:

```
----------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations      Bytes      Lines     Tokens
----------------------------------------------------------------------------------------------------------------------------
BM_CompileAPIFileDenseDecls<Phase::Check>/256       1077198 ns      1076381 ns         1024 4.89325M/s 181.163k/s 1.02287M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/1024      2017288 ns      2015795 ns         1024 16.1172M/s 485.665k/s 2.85347M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/4096      5578900 ns      5574719 ns          256 25.4511M/s 722.548k/s 4.27986M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/16384    20038101 ns     20024204 ns           64 30.5106M/s 814.664k/s  4.8337M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/65536    81702279 ns     81648274 ns           16 30.9184M/s 801.732k/s 4.75955M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/262144  349500301 ns    349287794 ns            8 29.2591M/s 750.467k/s 4.45562M/s
```

Note, benchmarks on my machine are expected to be noisy. But it does look like an incremental performance decrease in this test case.

### CODE REVIEW

**Comment by jonmeow** [comment]
Note, I'll lean on @chandlerc to decide whether this is worth it. The reduction in memory locality might be a net loss, performance versus space.

**Comment by zygoloid** [comment]
Yeah, this is what I was worried about -- a small performance decrease seems to be visible here, but we can probably get some performance benefits by avoiding looking at the operands of instructions in cases where we only need the kind. It's hard to know if that'll be a net benefit or regression unless we do the work to only look at the kind where possible.

Can you try changing `InstStore::Is` and `InstStore::TryGetAs` to look at the kind directly instead of first grabbing the whole `Inst`, and see if that makes any difference? (I'm not expecting a big change there, but you never know.)

**Comment by jonmeow** [comment]
> Yeah, this is what I was worried about -- a small performance decrease seems to be visible here, but we can probably get some performance benefits by avoiding looking at the operands of instructions in cases where we only need the kind. It's hard to know if that'll be a net benefit or regression unless we do the work to only look at the kind where possible.
> 
> Can you try changing `InstStore::Is` and `InstStore::TryGetAs` to look at the kind directly instead of first grabbing the whole `Inst`, and see if that makes any difference? (I'm not expecting a big change there, but you never know.)

Should be done. It looks like a small improvement, but it's hard to tell.

Note, down this road, there's also similar methods on `TypeStore`. But those are harder to adjust quickly due to dependencies and the templating (since it's templated, whereas `TypeStore` currently relies on the cpp file for inst access methods)

File: toolchain/sem_ir/inst.h [hunk 1]
```diff
@@ -427,22 +451,32 @@ class InstStore {
   // Reserves space.
   auto Reserve(size_t size) -> void {
     loc_ids_.reserve(size);
-    values_.Reserve(size);
+    types_and_args_.Reserve(size);
```
Comment (danakj) [inline toolchain/sem_ir/inst.h:493]: Do you want to reserve for kinds_ too?

File: toolchain/sem_ir/inst.h [hunk 2]
```diff
@@ -104,6 +106,16 @@ struct InstLikeTypeInfo<InstCat> : InstLikeTypeInfoBase<InstCat> {
 template <typename T>
 concept InstLikeType = requires { sizeof(InstLikeTypeInfo<T>); };
 
+// Supports storage of an `Inst` object's `kind` separate from its other
+// members.
+struct InstTypeAndArgs {
+  explicit InstTypeAndArgs(Inst inst);
+
+  TypeId type_id;
+  int32_t arg0;
```
Comment (danakj) [inline toolchain/sem_ir/inst.h:115]: aside: maybe these and the fields in Inst could be `AnyRawId` now, I didn't see them when I introduced it for the inst category types.

**Comment by jonmeow** [comment]
@danakj To be sure, what we're looking at right now is the possibility of a significant enough performance regression that I would just close this PR.

File: toolchain/sem_ir/inst.h (hunk 2 above)
Comment (jonmeow) [reply toolchain/sem_ir/inst.h:115]: I think the contract may differ. `AnyRawId` seems intended to represent an `Id` type, but we allow things such as  `BoolValue` on `BoolLiteral`.

Comment (jonmeow) [reply toolchain/sem_ir/inst.h:115]: Or maybe `BoolValue` is an `Id` too? That's a bit opaque.

Either way, I'd like to focus more on performance issues here.

File: toolchain/sem_ir/inst.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/sem_ir/inst.h:493]: Done, but this is only called once to reserve space for builtins, so it will be negligible for performance.

**Comment by jonmeow** [comment]
Flipping a CHECK to DCHECK in `As` (on both trunk and branch for this comparison):

branch:

```
----------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations      Bytes      Lines     Tokens
----------------------------------------------------------------------------------------------------------------------------
BM_CompileAPIFileDenseDecls<Phase::Check>/256       1071900 ns      1071338 ns         1024 4.91628M/s 182.015k/s 1.02769M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/1024      1971624 ns      1969992 ns         1024 16.4919M/s 496.956k/s 2.91981M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/4096      5489374 ns      5484396 ns          256 25.8703M/s 734.447k/s 4.35034M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/16384    20271944 ns     20249973 ns           64 30.1704M/s 805.581k/s 4.77981M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/65536    82548931 ns     82498951 ns           16 30.5996M/s 793.465k/s 4.71047M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/262144  357791786 ns    357549384 ns            8 28.5831M/s 733.127k/s 4.35267M/s
```

trunk:

```
----------------------------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations      Bytes      Lines     Tokens
----------------------------------------------------------------------------------------------------------------------------
BM_CompileAPIFileDenseDecls<Phase::Check>/256       1082382 ns      1081691 ns         1024 4.86923M/s 180.273k/s 1.01785M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/1024      1967895 ns      1966100 ns         1024 16.5246M/s  497.94k/s 2.92559M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/4096      5382641 ns      5379362 ns          256 26.3754M/s 748.788k/s 4.43528M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/16384    19875072 ns     19860513 ns           64  30.762M/s 821.379k/s 4.87354M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/65536    81334970 ns     81277123 ns           16 31.0596M/s 805.393k/s 4.78128M/s
BM_CompileAPIFileDenseDecls<Phase::Check>/262144  351240852 ns    350994095 ns            8 29.1169M/s 746.819k/s 4.43396M/s
```

This still looks like a 1-2% regression

**Comment by jonmeow** [comment]
I tried refactoring `GetAs` to avoid the full `Get` conversion, and it still seems not to affect performance. My leaning is now that this is going to remain a small hit to performance.

Looking at eval.cpp, I think it'd be hard to actually refactor things to avoid fetches in that case, and I suspect that'll be generally true. Closing, can reopen if someone has more ideas.

**Pull request closed without merging**